### DISCUSSION TITLE
Custom Row Class Example

### DISCUSSION THREAD

**Discussion by Blackstareye** [simolus3/drift#3120] [Q&A] [answered]
I love the idea that you can inject custom row classes to drift.

I tried the example on the doc page https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/

but unfortunately it doesnt work as expected

either the builder fails with:
```log
[WARNING] drift_dev on lib/models/database/create/create_db.drift:
line 6, column 61 of package:daily_moments/models/database/create/user.dart: Unexpected parameter birthday which has no matching column.
  ╷
6 │   User({required this.id, required this.name, required this.birthday});
```

or if i delete the database.g.dart before i get a broken database.g.dart
which has not the birthdate element in there
```dart
  @override
  Set<GeneratedColumn> get $primaryKey => {id};
  @override
  User map(Map<String, dynamic> data, {String? tablePrefix}) {
    final effectivePrefix = tablePrefix != null ? '$tablePrefix.' : '';
    return User( //<--- missing birthdate
      id: attachedDatabase.typeMapping
          .read(DriftSqlType.int, data['${effectivePrefix}id'])!,
      name: attachedDatabase.typeMapping
          .read(DriftSqlType.string, data['${effectivePrefix}name'])!,
    );
  }
```

DatabaseClass:
```dart
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:drift_flutter/drift_flutter.dart';
import 'package:path/path.dart' as p;


import 'package:daily_moments/models/database/create/user.dart';
import 'package:path_provider/path_provider.dart';

part 'database.g.dart';


@DriftDatabase(
  include: {'create/create_db.drift'}
)
class AppDatabase extends _$AppDatabase {
  static String databaseName = 'dailyMoments';
  AppDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;

  static QueryExecutor _openConnection() {
    // `driftDatabase` from `package:drift_flutter` stores the database in
    // `getApplicationDocumentsDirectory()`.
    return driftDatabase(name: AppDatabase.databaseName);
  }
  static Future<void> deleteDatabase() async {
    // `driftDatabase` from `package:drift_flutter` stores the database in
    // `getApplicationDocumentsDirectory()`.
    Directory folder = await getApplicationDocumentsDirectory();
    print(folder.listSync());
    File dbFile = File(p.join(folder.path, AppDatabase.databaseName));
    await dbFile.delete();
  }
}

```

Drift file:
```sql
import 'user.dart'; -- or what the Dart file is called

CREATE TABLE users(
  id INTEGER NOT NULL PRIMARY KEY,
  name TEXT NOT NULL,
  birth_date DATETIME NOT NULL
) WITH User;


```
I also tried `TEXT` instead of DATETIME

Model:
```dart
class User {
  final int id;
  final String name;
  final DateTime birthday;

  User({required this.id, required this.name, required this.birthday});
}
```
<hr>

So first of all I have a question:

* can the files be placed in different folders and paths or do they have to be all in **one** folder ?

And the other question:
 * what is wrong in my code ? 
    i followed the doc (but implemented it in my app, so there are other dependencies as well)

**Comment by Blackstareye** [comment, marked as answer]
Okay I analyzed it further and made an flutter example of it.

~~What works: ~~
~~* use CreationModels like `Users`~~
~~* use drift files without Rowclass models~~

~~what does not work:~~
~~* use drift files with `WITH` statement and custom rowclasses~~

~~Can somebody explain this or is this a bug ?~~

I think I found the error. It was a typo with `birth_date`.

AND about the path ...

I think it fixes it if I import the models in the `database.dart` file, cause `database.g.dart` is a part from it, right ?

**Reply by Blackstareye** [reply]
Yep now everything works. 

Thanks for Rubberducking ^^`
Closed

**Reply by simolus3** [reply]
> can the files be placed in different folders and paths or do they have to be all in one folder ?

Yes! Imports in drift files and the `include` parameter are resolved like Dart import statements. You can use relative paths (including across folders) or `package:` URIs.

**Reply by Blackstareye** [reply]
Thanks, now everything works as expected.

But there was one thing that could be added to the Docs I think:

There is an internal transformation between snake_case and camelCase.
It is expected that the Custom Rowclasses uses camelCase.

But in the docs ([link](https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/#using-custom-classes)) there is a line which I misunderstood:

>Each constructor argument must have the name of a drift column (matching the getter name in the table definition) 

I interpreted that so that the Rowclass Model should also use snake case when the DDL defines it like that. 
That was another reason while the build failed or the `Compagnion`-Classes missed some attributes.

So maybe this should be mentioned as well ? What do you think @simolus3 ?
